### PR TITLE
Add Compound v2 lending supply and borrow

### PR DESCRIPTION
**Pull request by bizzyvinci**
Brief comments on the purpose of your changes:
Adding compound v2 supply and borrow
* cTokens: [query](https://dune.com/queries/1903614)
* Supply: [query](https://dune.com/queries/1903672)
    - [x] Mints (or deposit)
    - [x] redeem
    - [ ] liquidation
* Borrow: [query](https://dune.com/queries/1911107)
    - [x] borrow
    - [x] repay
    - [x] liquidation

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by bizzyvinci** [comment]
What's the purpose of adding liquidations to lending supply.

There are 2 scenarios in AAVE based on the [receiveAToken parameter](https://github.com/aave/protocol-v2/blob/master/contracts/protocol/lendingpool/LendingPool.sol#L422) in liquidateCall.

1. There's a withdrawal and hence less supply in the pool when `receiveAToken = False`
2. There's constant supply--if there's no fee--in the pool when `receiveAToken = True`, but the owner of the position changes from the borrower to the liquidator

There's no if statement to separate the 2 scenarios in [AAVE model](https://github.com/duneanalytics/spellbook/blob/main/models/aave/ethereum/aave_v2_ethereum_supply.sql#L66)

Now, back to compound. All liquidations are scenario 2 in which positions of collateral--cTokens--are transferred from the borrower to the liquidator. And adding liquidation to supply is complicated for 2 reasons
1. underlying asset amount is not provided in call and events
2. Not all cToken is transferred to liquidator as some goes into reserve.

For reference, [this section](https://github.com/compound-finance/compound-protocol/blob/master/contracts/CToken.sol#L793-L840) is where the collateral transfer occurs.

File: tests/compound/ethereum/compound_v2_borrow_test.sql [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+select * from (
+    select * from unit_test1
+    union
+    select * from unit_test2
+    union
+    select * from unit_test3
+    union
+    select * from unit_test4
+)
+where test = false
```
Comment (Hosuke) [inline tests/compound/ethereum/compound_v2_borrow_test.sql:55]: We can use `union all` for the detailed info.
```suggestion
select * from (
    select 'test1' as test_no, * from unit_test1
    union all
    select 'test2' as test_no, * from unit_test2
    union all
    select 'test3' as test_no, * from unit_test3
    union all
    select 'test4' as test_no, * from unit_test4
)
where test = false
```

**Review by Hosuke — CHANGES_REQUESTED** [review]
Hi @bizzyvinci 

We may need to fix the test first.
[v2 test sql query](https://dune.com/queries/1922843?d=1)
[dbt test info](https://github.com/duneanalytics/spellbook/actions/runs/3995917855/jobs/6855347144#step:10:21)

**Comment by bizzyvinci** [comment]
@Hosuke, done

**Comment by Hosuke** [comment]
In the next few commits, I will try to further update the sql to incremental views.

File: models/compound/ethereum/compound_ethereum_schema.yml [hunk 2]
```diff
@@ -97,4 +97,125 @@ models:
...
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - block_date
+            - version
+            - transaction_type
+            - token_address
+            - evt_tx_hash
+            - borrower
+            - repayer
```
Comment (Hosuke) [inline models/compound/ethereum/compound_ethereum_schema.yml:202]: Can you help me to define the `unique key` combination for `compound_v2_ethereum_borrow`?

File: models/compound/ethereum/compound_ethereum_schema.yml [hunk 3]
```diff
@@ -97,4 +97,125 @@ models:
...
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - block_date
+            - version
+            - transaction_type
+            - token_address
+            - evt_tx_hash
+            - depositor
+            - withdrawn_to
```
Comment (Hosuke) [inline models/compound/ethereum/compound_ethereum_schema.yml:139]: Also can you help me to define the `unique key` combination for `compound_v2_ethereum_supply`?
We can add uniqueness tests in the schema file.

**Review by Hosuke — CHANGES_REQUESTED** [review]
Hi @bizzyvinci, I try to define the unique key combination for borrow and supply model, as incremental merge based on the unique key. However I may get the unique key combination wrong.

Can you help me to define the unique key combination for these views?

Thanks.

**Comment by bizzyvinci** [comment]
@Hosuke, done.

There is another set of wizards, they don't write SQL queries, they send transactions to the blockchain. Running the following would show [this transaction](https://etherscan.io/tx/0x9fdf7b3c0846629f0d469cb86af117a97f73c63d4ad3caf46b119a06e7159a89) in which the same address borrowed the same token 12 times 🤯

```sql
select evt_tx_hash, borrower, contract_address, count(*)
from compound_v2_ethereum.cErc20_evt_Borrow
group by 1, 2, 3
order by 4 desc
limit 100
```

Using block_number and evt_index is always a safe unique key when they are available. You can add back the other keys if they'll help queries to run faster like indexing.

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @bizzyvinci

This PR adds one static table [`compound_v2_ethereum_ctokens`](https://github.com/duneanalytics/spellbook/pull/2503/files#diff-c7980b9ff32186dbd31ed104741bb221e97bf790d32fc490f36bf0b474a3acba), and two incremental views [[`compound_v2_ethereum_borrow.sql`](https://github.com/duneanalytics/spellbook/pull/2503/files#diff-505900ad0ee6e5931995e50a8f9b32910a0cf1b3d50121830825c3b5ae2af32e), [`compound_v2_ethereum_supply`](https://github.com/duneanalytics/spellbook/pull/2503/files#diff-adf4666fb8c26afafe82fb36c4dcffe4e5cad3cd4aa6d127e2da2e957153e042)]

I have reviewed the PR and here is the summary:
- Metadata & naming conventions look good ✅
- Config blocks match uniqueness tests in schema files ✅
- Tests added for the new models ✅
- Incremental filters added when joining large tables ✅
- Contract addresses are lowercased ✅

**Comment by jeff-dude** [comment]
CI failing due to irrevelant test from another merged PR. will try again shortly.

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**